### PR TITLE
subscription: Only restart payload when needed

### PR DESCRIPTION
**Pull request by M4rtinK**
Only restart the payload after registration/unregistration
when necessary - eq. when CDN is in use or when CDN overrides
another installation source (CDROM).

Do not restart the payload in cases when it is not needed,
such as when the URL installation source is in use.

To do this, move the payload restart logic to the subscription
helper thread, and make it conditional, as payload restart is
currently not needed during registration triggered by an
automated kickstart installation, only during manual interaction
in the GUI.

Resolves: rhbz#1861351

### CODE REVIEW

**Comment by M4rtinK** [comment]
jenkins, test this please

**Comment by M4rtinK** [comment]
jenkins, test this please

File: tests/nosetests/dd_tests/dd_test.py [hunk 1]
```diff
@@ -10,7 +10,9 @@
 
 from contextlib import contextmanager
 from collections import namedtuple
-from rpmfluff import SourceFile, GeneratedSourceFile, SimpleRpmBuild, expectedArch, make_elf
+from rpmfluff import SourceFile, GeneratedSourceFile, SimpleRpmBuild
+from rpmfluff.utils import expectedArch
+from rpmfluff.make import make_elf
```
Comment (jkonecny12) [inline tests/nosetests/dd_tests/dd_test.py:15]: Please remove this. We will pin old version for RHEL-8.

File: pyanaconda/ui/lib/subscription.py [hunk 2]
```diff
@@ -334,10 +358,24 @@ def unregister(payload, overridden_source_type, progress_callback=None, error_ca
...
+                switched_source = True
+
+            # If requested, also restart the payload if:
+            # - installation source switch occured
+            # - the current source is CDN, which can no longer be used
+            #   after unregistration, so we need to refresh the Source
+            #   and Software spokes
+            if restart_payload and (source_proxy.Type == SOURCE_TYPE_CDN or switched_source):
+                log.debug("subscription thread: restarting payload after unregistration")
+                log.debug("AA UNREG RSTART")
```
Comment (jkonecny12) [inline pyanaconda/ui/lib/subscription.py:377]: I guess you don't want to merge this message.

Comment (M4rtinK) [reply pyanaconda/ui/lib/subscription.py:377]: So one debugging message did sneak through, thanks for catching it, should be fixed now. :)

File: tests/nosetests/dd_tests/dd_test.py (hunk 1 above)
Comment (M4rtinK) [reply tests/nosetests/dd_tests/dd_test.py:15]: OK, I'll drop it before merging. :)

File: pyanaconda/ui/lib/subscription.py [hunk 3]
```diff
@@ -278,16 +295,23 @@ def register_and_subscribe(payload, progress_callback=None, error_callback=None)
...
-        switch_source(payload, SOURCE_TYPE_CDN)
+    if payload.type == PAYLOAD_TYPE_DNF:
+        if source_proxy.Type in SOURCE_TYPES_OVERRIDEN_BY_CDN:
+            log.debug("subscription thread: overriding current installation source by CDN")
+            switch_source(payload, SOURCE_TYPE_CDN)
+        # If requested, also restart the payload if CDN is the installation source
+        # The CDN either already was the installation source or we just switched to it.
+        if restart_payload and source_proxy.Type == SOURCE_TYPE_CDN:
+            log.debug("subscription thread: restarting payload after registration")
+            _restart_payload(payload)
```
[the remark this answers is not in the document]
Comment (M4rtinK) [reply pyanaconda/ui/lib/subscription.py]: renamed to `_do_payload_restart()`. :)

**Comment by M4rtinK** [comment]
Port to Fedora: #2842

**Pull request merged**